ksqldb-common/src/main/java/io/confluent/ksql/logging/processing/ProcessingLoggerFactoryImpl.java

```diff
@@ -18,36 +18,123 @@
+
+import io.confluent.ksql.util.MetricsTagsUtil;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.CumulativeSum;
 
 public class ProcessingLoggerFactoryImpl implements ProcessingLoggerFactory {
+  public static final String PROCESSING_LOG_ERROR_METRIC_NAME = "processing-error-total";
+  public static final String PROCESSING_LOG_METRICS_GROUP_NAME = "processing-diagnostic-metrics";
+  public static final String PROCESSING_LOG_METRIC_DESCRIPTION =
+      "The total number of errors emitted by the processing log.";
+
   private final ProcessingLogConfig config;
   private final StructuredLoggerFactory innerFactory;
+  private final Metrics metrics;
+  private final Map<String, String> metricsTags;
   private final BiFunction<ProcessingLogConfig, StructuredLogger, ProcessingLogger> loggerFactory;
+  private final Function<Metrics, BiFunction<ProcessingLogger, Sensor, ProcessingLogger>>
+      loggerWithMetricsFactory;
+  private final Map<String, ProcessingLogger> processingLoggers;
 
   ProcessingLoggerFactoryImpl(
       final ProcessingLogConfig config,
-      final StructuredLoggerFactory innerFactory) {
-    this(config, innerFactory, ProcessingLoggerImpl::new);
+      final StructuredLoggerFactory innerFactory,
+      final Metrics metrics,
+      final Map<String, String> metricsTags
+  ) {
+    this(
+        config,
+        innerFactory,
+        metrics,
+        ProcessingLoggerImpl::new,
+        metricObject -> (processingLogger, sensor) ->
+            new MeteredProcessingLogger(processingLogger, metricObject, sensor),
+        metricsTags
+    );
   }
 
   ProcessingLoggerFactoryImpl(
       final ProcessingLogConfig config,
       final StructuredLoggerFactory innerFactory,
-      final BiFunction<ProcessingLogConfig, StructuredLogger, ProcessingLogger> loggerFactory
+      final Metrics metrics,
+      final BiFunction<ProcessingLogConfig, StructuredLogger, ProcessingLogger> loggerFactory,
+      final Function<Metrics, BiFunction<ProcessingLogger, Sensor, ProcessingLogger>>
+          loggerWithMetricsFactory,
+      final Map<String, String> metricsTags
   ) {
     this.config = config;
     this.innerFactory = innerFactory;
+    this.metrics = metrics;
     this.loggerFactory = loggerFactory;
+    this.loggerWithMetricsFactory = loggerWithMetricsFactory;
+    this.metricsTags = metricsTags;
+    this.processingLoggers = new ConcurrentHashMap<>();
   }
 
   @Override
-  public ProcessingLogger getLogger(final String name) {
-    return loggerFactory.apply(config, innerFactory.getLogger(name));
+  public ProcessingLogger getLoggerWithMetrics(
+      final String name
+  ) {
+    return getLoggerWithMetrics(name, "");
+  }
+
+  @Override
+  public ProcessingLogger getLoggerWithMetrics(
+      final String name,
+      final String queryId
+  ) {
+    if (processingLoggers.containsKey(name)) {
```

Comment:
Use computeIfAbsent here, or synchronize this method. There is still a race condition here where 2 threads can wind up creating a logger for a given name if they hit line 94 concurrently.

Reply:
I decided to go with adding the synchronized keyword